### PR TITLE
Add support for proxying additional backend types

### PR DESCRIPTION
**Pull request by iliajie**
Hey Jamie!

As Joe pointed out earlier, we need to support different backend types, particularly Unix sockets. This PR implements support for that.

### CODE REVIEW

**Comment by jcameron** [comment]
This could be cool, but let's hold off on new features until we do a new release of Virtualmin.

Also, it seems dangerous to allow non-root users to create Unix socket proxies.

**Comment by iliajie** [comment]
> This could be cool, but let's hold off on new features until we do a new release of Virtualmin.

This is a tiny change rather than a new big feature. Though wanted!
 
> Also, it seems dangerous to allow non-root users to create Unix socket proxies.

I'm not sure what you mean, but Unix sockets are much more secure, as they aren't available over a network and can only be used by processes with the right permissions.

So, why is this dangerous? What can a Unix socket proxy do that a TCP/IP socket proxy cannot?

**Comment by jcameron** [comment]
What I mean is, do we really want a domain owner to be able to configured Apache to connect to unix sockets? And even if we do, I think that this would be a very niche feature that the typical domain owner can't understand anyway. So we should limit it to root for now..

**Comment by iliajie** [comment]
> So we should limit it to root for now

I have fixed that, alongside with another bug.

Although, perhaps I could improve it further. I can parse the socket file path, check if it exists, and if it does, verify if it is owned by the domain owner user and/or located under the user's home directory?

**Comment by iliajie** [comment]
> > So we should limit it to root for now
> 
> I have fixed that, alongside with another bug.
> 
> Although, perhaps I could improve it further. I can parse the socket file path, check if it exists, and if it does, verify if it is owned by the domain owner user and/or located under the user's home directory?

Check this patch https://github.com/virtualmin/virtualmin-gpl/pull/825/commits/89dab8348431e01f7379ec3cca0b2731562e8a23. It will restrict what the domain user can do with the Unix socket file.

**Comment by jcameron** [comment]
Let's just limit this to root instead of adding permission checks ... I don't want to complicate the UI too much with a new socket type.

**Comment by iliajie** [comment]
> Let's just limit this to root instead of adding permission checks ... I don't want to complicate the UI too much with a new socket type.

Done.

File: edit_balancer.cgi [hunk 1]
```diff
@@ -53,20 +53,22 @@ if (($in{'new'} || !$b->{'balancer'}) &&
 		        ".forEach(function(radio) { radio.disabled = ".
 			    "this.checked; }, this);'");
 	}
-
+my $placeholder = "$text{'index_global_eg'} http://127.0.0.1:12345";
+$placeholder .= " $text{'or'} unix:/path/to/socket|http://"
```
Comment (jcameron) [inline edit_balancer.cgi:57]: What's this http:// doing at the end of the socket path?

Comment (iliajie) [reply edit_balancer.cgi]: > What's this http:// doing at the end of the socket path?

Thanks for pointing this out, as this part should be `http://127.0.0.1` instead!

Fixed now!

Comment (jcameron) [reply edit_balancer.cgi]: Do we really need to require that the user enter this though? If it's needed for unix domain sockets, we should add it automatically.

Comment (iliajie) [reply edit_balancer.cgi]: Require? No, we don't require it. This is a placeholder, a tip for the user what can be entered to the field potentially ..

Comment (jcameron) [reply edit_balancer.cgi]: What I mean is, in Apache to proxy to a unix domain socket that `|http://127.0.0.1` suffix is needed in ProxyPass for some reason, even though that's not the address the proxying is to. We should make things easier for users by adding and removing it automatically.

Comment (iliajie) [reply edit_balancer.cgi]: I don't think we can do this, as `http://127.0.0.1` in UDS isn't a static value.

Comment (jcameron) [reply edit_balancer.cgi]: What is that URL used for then?

Comment (iliajie) [reply edit_balancer.cgi]: > What is that URL used for then?

It's used to define the communication backend protocol, which in my view can be anything, like `|wss://12.33.44.55` or `|uwsgi://localhost`.

Comment (jcameron) [reply edit_balancer.cgi]: Ah so the IP address in the URL doesn't matter, only the protocol like ws or uwsgi ?

Comment (iliajie) [reply edit_balancer.cgi]: > Ah so the IP address in the URL doesn't matter, only the protocol like ws or uwsgi ?

The protocol and IP matters.

Comment (jcameron) [reply edit_balancer.cgi]: So from looking at the docs, it seems like only the protocol and path matters : https://httpd.apache.org/docs/current/mod/mod_proxy.html#proxypass

What would the IP/host mean when the connection is always being made to a socket?

Comment (jcameron) [reply edit_balancer.cgi]: What I'm really wondering is if there's a more user-friendly way to allow users to proxy to unix domain sockets than expecting them to know this rather confusing Apache syntax...

Comment (jcameron) [reply edit_balancer.cgi]: Anyway, let's not hold up this PR on UI details ... we can address that later.

Comment (iliajie) [reply edit_balancer.cgi]: > Anyway, let's not hold up this PR on UI details ... we can address that later.

Great!

**Pull request merged**